testutil/e2e/e2e.go

```diff
@@ -776,6 +782,324 @@ func (lt *lavaTest) checkPayments(testDuration time.Duration) {
+					panic("PROVIDER PAYMENT CHECK FAILED")
+				}
+			} else {
+				netAmount := balanceRes.GetBalance().Amount.Int64()
+				if netAmount < providerBalances[provider].GetBalance().Amount.Int64() {
+					panic("PROVIDER PAYMENT CHECK FAILED")
+				}
+			}
+		}
+	}
+	utils.LavaFormatInfo("PROVIDER BALANCE CHECK AFTER PAYMENT OK")
+}
+
+func (lt *lavaTest) checkQoS() error {
+	utils.LavaFormatInfo("Starting QoS Tests")
+	errors := []string{}
+
+	pairingClient := pairingTypes.NewQueryClient(lt.grpcConn)
+	providerCU, err := calculateProviderCU(pairingClient)
+	if err != nil {
+		panic("Provider CU calculation error!")
+	}
+
+	providerIdx := 0
+	for provider := range providerCU {
+		// Get sequence number of provider
+		logNameAcc := "8_authAccount" + fmt.Sprintf("%02d", providerIdx)
+		lt.logs[logNameAcc] = new(bytes.Buffer)
+
+		fetchAccCommand := lt.lavadPath + " query account " + provider + " --output=json"
+		cmdAcc := exec.CommandContext(context.Background(), "", "")
+		cmdAcc.Path = lt.lavadPath
+		cmdAcc.Args = strings.Split(fetchAccCommand, " ")
+		cmdAcc.Stdout = lt.logs[logNameAcc]
+		cmdAcc.Stderr = lt.logs[logNameAcc]
+		err = cmdAcc.Start()
+		if err != nil {
+			errors = append(errors, fmt.Sprintf("%s", err))
+		}
+		lt.commands[logNameAcc] = cmdAcc
+		err = cmdAcc.Wait()
+		if err != nil {
+			errors = append(errors, fmt.Sprintf("%s", err))
+		}
+
+		var obj map[string]interface{}
+		err := json.Unmarshal((lt.logs[logNameAcc].Bytes()), &obj)
+		if err != nil {
+			panic(err)
+		}
+
+		sequence, ok := obj["sequence"].(string)
+		if !ok {
+			panic("Sequence field is not valid!")
+		}
+		sequenceInt, err := strconv.ParseInt(sequence, 10, 64)
+		if err != nil {
+			panic(err)
+		}
+		sequenceInt--
+		sequence = strconv.Itoa(int(sequenceInt))
+		//
+		logName := "9_QoS_" + fmt.Sprintf("%02d", providerIdx)
+		lt.logs[logName] = new(bytes.Buffer)
+
+		txQueryCommand := lt.lavadPath + " query tx --type=acc_seq " + provider + "/" + sequence
+
+		cmd := exec.CommandContext(context.Background(), "", "")
+		cmd.Path = lt.lavadPath
+		cmd.Args = strings.Split(txQueryCommand, " ")
+		cmd.Stdout = lt.logs[logName]
+		cmd.Stderr = lt.logs[logName]
+
+		err = cmd.Start()
+		if err != nil {
+			errors = append(errors, fmt.Sprintf("%s", err))
+		}
+		lt.commands[logName] = cmd
+		err = cmd.Wait()
+		if err != nil {
+			errors = append(errors, fmt.Sprintf("%s", err))
+		}
+
+		lines := strings.Split(lt.logs[logName].String(), "\n")
+		for idx, line := range lines {
+			if strings.Contains(line, "key: QoSScore") {
+				startIndex := strings.Index(lines[idx+1], "\"") + 1
+				endIndex := strings.LastIndex(lines[idx+1], "\"")
+				qosScoreStr := lines[idx+1][startIndex:endIndex]
+				qosScore, err := strconv.ParseFloat(qosScoreStr, 64)
+				if err != nil {
+					errors = append(errors, fmt.Sprintf("%s", err))
+				}
+				if qosScore < 1 {
+					errors = append(errors, "QoS score is less than 1 !")
```

Comment:
Can you also print what is the quality of service received? so it will be easier to validate?